Microsoft.Toolkit.HighPerformance/Memory/Span2D{T}.cs

```diff
@@ -0,0 +1,1056 @@
+        /// </summary>
+        /// <param name="i">The target row to get the element from.</param>
+        /// <param name="j">The target column to get the element from.</param>
+        /// <returns>A reference to the element at the specified indices.</returns>
+        [Pure]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe ref T DangerousGetReferenceAt(int i, int j)
+        {
+#if SPAN_RUNTIME_SUPPORT
+            ref T r0 = ref MemoryMarshal.GetReference(this.span);
+#else
+            ref T r0 = ref RuntimeHelpers.GetObjectDataAtOffsetOrPointerReference<T>(this.Instance, this.Offset);
+#endif
+            int index = (i * (this.width + this.Pitch)) + j;
+
+            return ref Unsafe.Add(ref r0, (IntPtr)(void*)(uint)index);
+        }
+
+        /// <summary>
+        /// Slices the current instance with the specified parameters.
+        /// </summary>
+        /// <param name="row">The target row to map within the current instance.</param>
+        /// <param name="column">The target column to map within the current instance.</param>
+        /// <param name="height">The height to map within the current instance.</param>
+        /// <param name="width">The width to map within the current instance.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when either <paramref name="height"/>, <paramref name="width"/> or <paramref name="height"/>
+        /// are negative or not within the bounds that are valid for the current instance.
+        /// </exception>
+        /// <returns>A new <see cref="Span2D{T}"/> instance representing a slice of the current one.</returns>
+        [Pure]
+        public Span2D<T> Slice(int row, int column, int height, int width)
+        {
+            if ((uint)row >= Height)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeExceptionForRow();
+            }
+
+            if ((uint)column >= this.width)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeExceptionForColumn();
+            }
+
+            if ((uint)height > (Height - row))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeExceptionForHeight();
+            }
+
+            if ((uint)width > (this.width - column))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeExceptionForWidth();
+            }
+
+            int
+                shift = ((this.width + this.Pitch) * row) + column,
+                pitch = this.Pitch + (this.width - width);
+
+#if SPAN_RUNTIME_SUPPORT
+            unsafe
+            {
+                ref T r0 = ref Unsafe.Add(ref MemoryMarshal.GetReference(this.span), (IntPtr)(void*)(uint)shift);
+
+                return new Span2D<T>(ref r0, height, width, pitch);
+            }
+#else
+            IntPtr offset = this.Offset + (shift * Unsafe.SizeOf<T>());
+
+            return new Span2D<T>(this.Instance!, offset, height, width, pitch);
```

Comment:
What exactly guarantees `Instance` to be non-null here?

Reply:
Good catch! I was actually just about to update this exact line 😄

That was guaranteed by the fact that the size was being checked before, so the target instance wouldn't have been `null` then. But that was before I added support for `void*` based 2D spans on .NET Standard <= 2.0 too, so that's no longer the case now, and `null` is possible and valid in this case. I'll go update this! 👍